### ISSUE TITLE
Support torch backend in optimization module

### ISSUE THREAD

**Issue by drpaprika** [HarrisonKramer/optiland#220]
Hi all,
At the moment `torch` is not supported in the optimization module because `scipy` only supports `numpy` arrays:

```python3
if be.get_backend() != "numpy":
    raise ValueError(
        "Optiland optimization currently only supports NumPy backend."
    )
```

I believe that supporting PyTorch for optimization could yield speed improvements as the optical system is traced repeatedly during merit function evaluation and optimization, and Pytorch offers powerfull optimizers like Adam.

At first, a simple implementation could be like the following:

```python3
def optimize(self, method=None, maxiter=1000, disp=True, tol=1e-3, callback=None):

    x0 = [var.value for var in self.problem.variables]
    self._x.append(x0)

    if be.backend_name == "numpy":
        x0_np = be.to_numpy(x0)
        bounds = tuple([var.bounds for var in self.problem.variables])
        options = {"maxiter": maxiter, "disp": disp}

        with warnings.catch_warnings():
            warnings.simplefilter("ignore", category=RuntimeWarning)
            result = optimize.minimize(
                self._fun,
                x0_np,
                method=method,
                bounds=bounds,
                options=options,
                tol=tol,
                callback=callback,
            )
    
    elif be.backend_name == "torch":
        # Create torch tensor for variables, requires gradients
        x = be.tensor(x0, dtype=be.precision, device=be.device, requires_grad=True)

        # Choose optimizer
        optimizer = be.optim.LBFGS([x], max_iter=maxiter, tolerance_grad=tol)

        def closure():
            optimizer.zero_grad()
            loss = self._fun(x)
            loss.backward()
            return loss

        # Perform optimization
        optimizer.step(closure)

        # Build a dummy result object similar to scipy.optimize.OptimizeResult
        result = {
            "x": x.detach().cpu().numpy(),
            "fun": self._fun(x).item(),
            "success": True,
            "message": "Optimized with torch.optim.LBFGS",
        }

    else:
        raise ValueError(
            f'"{be.name}" backend is not supported in optimization.'
        )

    # The last function evaluation is not necessarily the lowest.
    # Update all lens variables to their optimized values
    for idvar, var in enumerate(self.problem.variables):
        var.update(x[idvar])
    self.problem.update_optics()

    return result
```

Also the arrays need to be detached to cpu to work with pandas. If the values are torch tensors on a GPU or with `requires_grad=True`, PyTorch raises:

- `On GPU: "can't convert cuda:0 device type tensor to numpy"`
- `With gradients: "Can't call numpy() on Tensor that requires grad"`

Unfortunately I'm not familliar with `torch` and is implementation in optiland's backend. I opened this issue to initiate the discussion on this topic and see where I can contribute, even though a PyTorch user would for sure propose a better implementation. We can also add this topic in the readme to seek contributors.

Cheers,
drpaprika

**Comment by HarrisonKramer** [comment]
Hi @drpaprika,

Thanks for opening this! You're right. Supporting PyTorch here could offer big advantages.

We had started this discussion when we were first rolling out the PyTorch backend. At the time, we weren't sure how best to handle the optimization in a clean way, so we held off on defining it for the torch backend. That means you must optimize "semi-manually", first retrieving the parameters you want to optimize. I am open to discussion here, but I am personally leaning toward making the `OptimizationProblem` class backend-agnostic (it already mostly is), and then supporting backend-specific optimizer classes like `NumpyOptimizer` and `TorchOptimizer` (or similar). It would be nice to have a `OptimizationProblem.parameters()` method to retrieve the parameters to be optimized, following a similar API to that used in torch. I think that makes it more intuitive for users that are used to the torch way of working.

In general, the idea would be: you define your problem in the same way (adding operands, variables, etc.), and then choose the optimizer class based on the backend you want to use. We could even do this selection automatically, but I think giving users explicit control might be the better path. The torch version could then use standard PyTorch optimizers (e.g. Adam, LBFGS), but encapsulated inside a new `TorchOptimizer` (or similar) class that handles parameter syncing and variable updates behind the scenes.

This isn’t fully fleshed out yet. I'm definitely open to ideas on how to structure the parameter syncing and optimizer interface. If you or anyone else wants to prototype something, bounce thoughts around, or even just share what API structure you think makes intuitive sense, that’d be very welcome. I will give it some thought as well. I suppose my ultimate goal is to make the optimization framework relatively backend-agnostic while also exposing a "torch-like" API for those users who are more familiar with PyTorch than they are with Optiland.

It would be great to make Optiland behave like a differentiable physics module that just drops into the torch graph. That way, users could easily integrate physical optics constraints into a wide variety of learning pipelines, whether model-based design, inverse problems, generative models, etc.

Regards,
Kramer

**Comment by drpaprika** [comment]
Hi @HarrisonKramer,
Thank you for your feedback. I see that you have begun refactoring the optimization module, should we wait for it to initiate the development?

**Comment by HarrisonKramer** [comment]
Hi @drpaprika,

I gave this some more thought and wanted to share a few potential directions for the PyTorch integration. I think all of these could coexist and be made available to users, depending on their level of flexibility or abstraction.

1. Torch-Aware Optimizer for `OptimizationProblem`

   This is the approach you suggested: an optimizer takes in an `OptimizationProblem` configured for the torch backend, and automatically dispatches to a PyTorch optimizer (e.g., Adam, LBFGS). This provides a backend-agnostic interface for users who simply want to optimize a system without worrying about the internals.

2. Torch-Compatible `OptimizationProblem` with .parameters()

    This would extend `OptimizationProblem` with a `.parameters()` method that returns trainable `torch.nn.Parameter` instances (e.g., surface radii, thicknesses, etc.), allowing it to be used directly in standard torch optimization loops.

Pseudo-code example:
```python
problem = optimization.OptimizationProblem()
# ... define operands and variables ...

optimizer = torch.optim.Adam(problem.parameters(), lr=1e-2)

for step in range(100):
    optimizer.zero_grad()
    loss = compute_loss(lens)  # e.g., RMS spot size, Strehl ratio, or custom calculation
    loss.backward()
    optimizer.step()
```
This makes the integration feel native to PyTorch users while still leveraging Optiland's physics engine.

3. Torch `nn.Module` Wrapper for Full Integration

    We can also provide a higher-level interface via a subclassable `nn.Module` that wraps an `OptimizationProblem`. This enables plug-and-play use of Optiland as a differentiable component within larger deep learning pipelines (e.g. for inverse design, optimization, or generative models, etc.).

Pseudo-code:
```python
import torch
import torch.nn as nn

class OptimizationProblemModule(nn.Module):
    def __init__(self, problem: OptimizationProblem, lens: Optic):
        super().__init__()
        self.problem = problem
        self.lens = lens
        self.params = nn.ParameterList(problem.parameters())

    def forward(self):
        """Define a PSF calculation for forward pass"""
        psf_result = psf.FFTPSF(self.lens, field=(0, 0), wavelength=0.55)
        return psf_result.psf  # returns a 2D tensor
```

This can then be used in a generative or data-driven training loop:

```python
model = OptimizationProblemModule(problem, lens)
target_psf = torch.load("measured_psf.pt")

optimizer = torch.optim.Adam(model.parameters(), lr=1e-2)

for step in range(100):
    optimizer.zero_grad()
    predicted_psf = model()
    loss = torch.nn.functional.mse_loss(predicted_psf, target_psf)
    loss.backward()
    optimizer.step()
```

This structure is relatively compact and extensible, e.g., it could easily be integrated into a GAN or VAE architecture where the optics act as a generative component, or used in inverse problems where the physical parameters are learned directly from data.

Just a few directions to consider.. We don't have to commit to any specific path yet, but I think there's real potential to make Optiland feel like a native part of the torch ecosystem, especially for users who want to merge physics and machine learning more seamlessly.

Happy to explore more if interesting.

Kramer

**Comment by HarrisonKramer** [comment]
Torch optimization is now supported as of #301.

Documented examples can be found [here](https://optiland.readthedocs.io/en/latest/gallery/optimization.html).

Closing this as complete.
